code/modules/reagents/chemistry/machinery/reagentgrinder.dm

```diff
@@ -64,334 +48,475 @@
+	. += span_notice("Its maintainence panel can be [EXAMINE_HINT("screwed")] [panel_open ? "closed" : "open"]")
+	if(panel_open)
+		. += span_notice("It can be [EXAMINE_HINT("pried")] apart")
 
-/obj/machinery/reagentgrinder/attack_hand_secondary(mob/user, list/modifiers)
+/obj/machinery/reagentgrinder/update_overlays()
 	. = ..()
-	if(. == SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN)
-		return
-	if(operating || !user.can_perform_action(src, ALLOW_SILICON_REACH | FORBID_TELEKINESIS_REACH))
-		return
-	eject(user)
-	return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
 
-/obj/machinery/reagentgrinder/attack_robot_secondary(mob/user, list/modifiers)
-	return attack_hand_secondary(user, modifiers)
+	if(!QDELETED(beaker))
+		. += "[base_icon_state]-beaker"
 
-/obj/machinery/reagentgrinder/attack_ai_secondary(mob/user, list/modifiers)
-	return attack_hand_secondary(user, modifiers)
+	if(anchored && !panel_open && is_operational && !(machine_stat & (BROKEN | NOPOWER)))
+		. += "[base_icon_state]-on"
 
 /obj/machinery/reagentgrinder/Exited(atom/movable/gone, direction)
 	. = ..()
 	if(gone == beaker)
 		beaker = null
-		update_appearance()
-	if(holdingitems[gone])
-		holdingitems -= gone
-
-/obj/machinery/reagentgrinder/proc/drop_all_items()
-	for(var/i in holdingitems)
-		var/atom/movable/AM = i
-		AM.forceMove(drop_location())
-	holdingitems = list()
-
-/obj/machinery/reagentgrinder/update_icon_state()
-	icon_state = "[base_icon_state][beaker ? 1 : 0]"
-	return ..()
+		update_appearance(UPDATE_OVERLAYS)
 
+/obj/machinery/reagentgrinder/RefreshParts()
+	. = ..()
+
+	for(var/datum/stock_part/part in component_parts)
+		if(istype(part, /datum/stock_part/servo))
+			speed = part.tier
+		else if(istype(part, /datum/stock_part/matter_bin))
+			maximum_weight = WEIGHT_CLASS_GIGANTIC * part.tier
+/**
+ * Inserts, removes or replaces the beaker present
+ * Arguments
+ *
+ * * mob/living/user - the player performing the action
+ * * obj/item/reagent_containers/new_beaker - the new beaker to replace the old, null to do nothing
+ */
 /obj/machinery/reagentgrinder/proc/replace_beaker(mob/living/user, obj/item/reagent_containers/new_beaker)
-	if(!user)
-		return FALSE
-	if(beaker)
+	PRIVATE_PROC(TRUE)
+
+	if(!QDELETED(beaker))
 		try_put_in_hand(beaker, user)
-		beaker = null
-	if(new_beaker)
+
+	if(!QDELETED(new_beaker))
+		if(!user.transferItemToLoc(new_beaker, src))
+			return
 		beaker = new_beaker
-	update_appearance()
-	return TRUE
+
+	update_appearance(UPDATE_OVERLAYS)
+
+/**
+ * Transfer this item or items inside if its a bag into the grinder
+ * Arguments
+ *
+ * * mob/user - the player who is inserting these items
+ * * list/obj/item/to_add - list of items to add
+ */
+/obj/machinery/reagentgrinder/proc/load_items(mob/user, list/obj/item/to_add)
+	PRIVATE_PROC(TRUE)
+
+	//surface level checks to filter out items that can be grinded/juice
+	var/list/obj/item/filtered_list = list()
+	for(var/obj/item/weapon as anything in to_add)
```

Comment:
why is the var name weapon? hold over?

Reply:
its not a type just a var name. renamed weapon -> ingredient for clarity